src/index.tsx

```diff
@@ -51,31 +52,69 @@ class CommLock {
   promise: any;
   requestedLockCount: number;
   clearedLockCount: number;
-
-  constructor() {
+  _activeTimeout: number;
+  _lockTimeout: number;
+  _timeoutSignal: Signal<CommLock, number>;
+  _recentKernelState: string;
+  _sessionContext: ISessionContext;
+
+  constructor(_lockTimeout: number, sessionContext: ISessionContext) {
+    this._lockTimeout = _lockTimeout;
+    this._activeTimeout = null;
     this.requestedLockCount = 0;
     this.clearedLockCount = 0;
     this._disable = {};
     this.promise = { 0: Promise.resolve() };
+    this._timeoutSignal = new Signal<CommLock, number>(this);
+    this._sessionContext = sessionContext;
+    this._sessionContext.statusChanged.connect(this.kernelStateRecorder, this);
+    this._timeoutSignal.connect(this.timeoutExpireHandler, this);
+  }
+
+  kernelStateRecorder(sender: ISessionContext, args: Kernel.Status) {
+    this._recentKernelState = args;
+  }
+
+  _clearTimeout() {
+    window.clearTimeout(this._activeTimeout);
+    this._activeTimeout = null;
+  }
+
+  /*
+		If the kernel was busy the last time, we assume it was busy executing 
+		code and we restart the timeout.
+  */
+  timeoutExpireHandler(sender: CommLock, args: number) {
+    this._clearTimeout();
+    if (this._recentKernelState === 'busy') {
+      console.log('Extending Timeout For: ', args);
+      this.createTimeout(args);
+    } else {
+      this.disable(args);
+    }
+  }
+
+  createTimeout(id: number) {
+    this._activeTimeout = setTimeout(() => {
+      this._timeoutSignal.emit(id);
+    }, this._lockTimeout);
   }
 
   enable(id: number): void {
     this.requestedLockCount++;
     this.promise[id] = new Promise(resolve => {
       this._disable[id] = resolve;
-      // Timeout and release the lock 1.5 sec after previous lock was released
-      setTimeout(() => {
-        if (this._disable?.[id]) {
-          this._disable[id]?.();
-        }
-      }, 1500 * (this.requestedLockCount - this.clearedLockCount));
     });
   }
 
-  disable(): void {
+  disable(id: number): void {
     this.clearedLockCount++;
-    this._disable[this.clearedLockCount]?.();
-    delete this._disable[this.clearedLockCount];
+    this._disable[id]?.();
+    delete this._disable[id];
+    this._clearTimeout();
+    if (this.clearedLockCount < this.requestedLockCount) {
```

Comment:
`id` and `clearedLockCount` seem both independent and dependent on each other. `clearedLockCount` is not set to the last `id` we got (but is monotonically increasing) while we compare it to how many locks we may want. If we had a bug and got a double message, this relationship would have hard to debug side effects.

Can we encode this strictly increasing relationship a bit better for id?

Reply:
It is set to the last received id. This should fix it.